README.md
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/1aa1a2fafa1a41e1bdcba6e453fcba1f)](https://app.codacy.com/app/DigitalCoin1/LightBurden?utm_source=github.com&utm_medium=referral&utm_content=DigitalCoin1/LightBurden&utm_campaign=Badge_Grade_Dashboard)
[![Build Status](https://travis-ci.org/DigitalCoin1/LightBurden.svg?branch=master)](https://travis-ci.org/DigitalCoin1/LightBurden)
<br>
[![N|Solid](https://i.imgur.com/0BOgWZy.png)](https://lightburden.org/)
# Lightburden - Sustainability in technology
# Version: 2.6.5.4
Lightburden came with the mission to renew the world of the exhaustive consumption of electric energy.

With the PoS mining concept, we have an ecologically correct coin, since it is not necessary to use powerful mining machines to obtain rewards for solved blocks.

Also embrace this idea and help build a cleaner world, we count on your help!

All the balance that is collected by Spero is aimed directly at the growth of the structuring of the currency and the social programs that it covers.

# Contact:
> :speech_balloon: Telegram:
**https://t.me/joinchat/G7cebkL3dbRqfuy1gVNm3g**

> :speech_balloon: Slack:
**https://lightburden.slack.com**

> :speech_balloon: Facebook:
**https://www.facebook.com/lightburdenofficial/**

> :speech_balloon: Twitter:
**https://twitter.com/Spero_Official**

> :speech_balloon: BTCTalk:
**https://bitcointalk.org/index.php?topic=2260985.0**

> :email: E-mail:
**lightburden@gmail.com**

> :earth_americas: Website:
**https://lightburden.org**

> :earth_americas: Block Explorer:
**https://explorer.lightburden.org**

> :droplet: Faucet:
**https://faucet.lightburden.org**

> :busts_in_silhouette: Support:
**https://suporte.lightburden.org**

# Specifications:
  - **ALGORITHM:** x13
  - **BLOCK TIME:** 60 seconds
  - **AWARDS PoW:** 2 LBDN to block 263250
  - **PoS:** After 2h
  - **PoS:** 25% per year
  - **CONFIRMATIONS:** 5 for mining and 3 for transactions
  - **E2M MESSAGE** Send and receive encrypted messages directly into your wallet
  - **TOTAL COINS:** ~ 7 million
  - **LAST BLOCK PoW:** 33331
  - **PRE-MINED COINS:** ~ 100000 LBDN
  - **VELOCITY:** Triple verification system in every network transaction, thus preventing the "double spend"
  - **Adaptable Block Size(ABS):** The block holds all transactions sent within 60 seconds, regardless of the number of transactions


# Phases of Lightburden (Blocks):

  - Proof of Work + Proof of Stake: 0 - 33331
  - Proof of Stake: 33331 - 263250
  - Proof of Work + Proof of Stake: acima de 263251

# Compile Source<br>
**Download and install the dependencies:**<br>
sudo apt-get install build-essential libboost-all-dev libcurl4-openssl-dev libdb5.3-dev libdb5.3++-dev libminiupnpc-dev qrencode libqrencode-dev git libtool automake autotools-dev autoconf pkg-config libssl-dev libgmp3-dev libevent-dev bsdmainutils
<br><br><br>
**Compile the daemon in the LightBurden/src directory:**<br>
cd LightBurden/src<br><br>
make -f makefile.unix USE_UPNP=1 USE_IPV6=1<br><br>
strip LightBurdend<br>
<br><br>
**Run daemon in the LightBurden/src directory:**<br>
./LightBurdend<br><br>
<br>
**(OPTIONAL)Compile the QT in the LightBurden directory:**<br>
sudo apt-get install libqt5gui5 libqt5core5a libqt5dbus5 qttools5-dev qttools5-dev-tools libprotobuf-dev protobuf-compiler libqrencode-dev qt5-default<br><br>
qmake LightBurden-qt.pro "USE_UPNP=1" "USE_QRCODE=1"<br><br>
make -f Makefile<br><br>

# lightburden.conf
```sh
listen=1
txindex = 1
rpcuser=user
rpcpassword=Random_Password
rpcport=55681
port=55680
#(0=off, 1=on) daemon - run in the background as a daemon and accept commands
daemon=1
#(0=off, 1=on) server - accept command line and JSON-RPC commands
server=1
rpcallowip=127.0.0.1
testnet=0
addnode=seed1.lightburden.org:55680
addnode=seed2.lightburden.org:55680
addnode=seed3.lightburden.org:55680
addnode=seed4.lightburden.org:55680

```

# lightburden.conf - Testnet
```sh
listen=1
txindex = 1
rpcuser=user
rpcpassword=Random_Password
rpcport=55691
port=55690
#(0=off, 1=on) daemon - run in the background as a daemon and accept commands
daemon=1
#(0=off, 1=on) server - accept command line and JSON-RPC commands
server=1
rpcallowip=127.0.0.1
testnet=1
addnode=seed1.lightburden.org:55690
addnode=seed2.lightburden.org:55690
addnode=seed3.lightburden.org:55690
addnode=seed4.lightburden.org:55690
```

# Exchanges:
 - 💎 https://exchange.lightburden.org
 - 💎 https://catalogop2p.com.br/lightburden-sustentabilidade-em-tecnologia

#  Mining:
- Solo Cpu:
```sh
{
"_comment1" : "Any long-format command line argument ",
"_comment2" : "may be used in this JSON configuration file",

"api-bind" : "127.0.0.1:4048",

"url" : "localhost:55681",
"user" : "user",
"pass" : "password",

"algo" : "x13",
"threads" : 0,
"cpu-priority" : 0,
"cpu-affinity" : -1,

"benchmark" : false,
"debug" : true,
"protocol": false,
"quiet" : false
}
```

```sh
cpuminer -q --algo=x13 -o 127.0.0.1:55681 -u user-p password
```

- Solo GPU:
```sh
ccminer30.exe -q --algo=x13 -o 127.0.0.1:55681 -u user-p password
```

# Timeline:
https://github.com/DigitalCoin1/LightBurden/blob/master/changelog.md


changelog.md
>$0001 2017/09/26<br>
Initial Upload to GitHub<br>

>$0006 2017/10/01<br>
In this update, the LightBurden-qt.pro file has been changed for compilation in the Linux environment.<br>

>$0007 2017/10/06<br>
Update README.md and add files: src/leveldb/Makefile<br>

>$0008 2017/10/08<br>
Update infos about end phase of PoW and start phase PoS;<br>
Update font of qt wallet;<br>
Update version;<br>
Update Protocol Others general's updates<br>

>$0010 2017/10/10<br>
Update for new implementations:<br>
→ Added option of cleaning of orphans blocks;<br>
→ Added display option "STAKE" in the "type" column inside the "Transactions" tab;<br>
→ Added "setban" node "" add | remove "(bantime)" / "listbanned" and "clearbanned" in the RPC console;<br>

>$0012 2017/10/11<br>
Ban Forked Peers who are stuck in a Getblocks Loop.<br>

>$0013 2017/10/22<br>
→ Update of strings names;<br>
→ Updating information in README.md<br>
→ Update Stake Info in QT  Wallet<br>

>$0015 2017/11/18<br>
→ Release of testnet mining;<br>
→ File Fix:<br>
   • main.cpp<br>
   • alter.cpp<br>
   • checkpoints.cpp<br>
   • checkpoints.h<br>

>$0016 2017/11/30<br>
Add node in the source:<br>
addnode=http://dnslightburden.ddnsking.com:55680<br>

>$0017 2018/03/28<br>
→ Add RPC Command "GETWALLETINFO"<br>
→ Update of transaction maturity icons<br>
→ splash screen main update<br>
→ Updating the logo inside the Wallet<br>
→ Added hashGenesisBlockTestNet = 0x00001fe23e7d6951a2a07ad684ff208fab9110fad607fabd81482a1692382f8f in main.h file<br>
→ Added testnet block.nNonce in main.cpp file (block.nNonce = 15547;)<br>
→ Added static MapCheckpoints mapCheckpointsTestnet = (0, uint256("0x000006f4925ed7b889f847f24621390d4943466f091c3254ca1bd8becc517f7b")) in the checkpoints.cpp file<br>
→ Change the initial letter of the testnet addresses in the base58.h file (PUBKEY_ADDRESS_TEST = 125, // Initial address letter (63 = s) -TESTNET))<br>
→ Change the example initial letter in the file qt / sendcoinsdialog.cpp<br>
(ui-> lineEditCoinControlChange-> setPlaceholderText (tr ("Enter the LightBurden address (e.g. Sjz75uKHzUQJnSdzvpiigEGxseKkDhQToX)"));)<br>
→ Changed version to 2.6.4.3<br>

>$0019 2018/07/15<br>
→ Added RPC command "getnetworkinfo" - Change in files: bitcoinrpc.cpp; bitcoinrpc.h; rpcblockchain.cpp; rpcnet.cpp<br>
→ Added RPC command "getblockchaininfo" - Change in files:bitcoinrpc.cpp; bitcoinrpc.h; rpcblockchain.cpp; rpcnet.cpp<br>
→ Added Wallet Status Information (Locked or Unlocked) in the RPC command "getinfo"<br>
→ Added new points in the network map updating the mapping up to block 130000- Change in file: checkpoints.cpp<br>
→ Changing copyrights, updating the start date of LightBurden works, being: "Copyright © 2022 The Lightburden developers" - Change in ".ts" files in the "src \ qt \ locale" folder<br>
→ Update of QT splashscreen images<br>
→ Changing QT fonts for "Century Gothic"<br>
→ Clear Orphans with automatic transaction page refresh, no need to restart wallet<br>
→ Changed version to 2.6.4.4<br>

>$0020 2018/07/31<br>
→ Add/Remove Nodes<br>

>$0021 2018/10/27<br>
→ Added new points in the network map updating the mapping up to block 165962- Change in file: checkpoints.cpp<br>
→ Performance Enhancements - Increased Speed ​​in Synchronization - Changes in files: main.cpp; main.h;<br>
→ Added new icons<br>
→ Change style of block download progress bar from "QWindowsXPStyle" to "QWindowsVistaStyle" - Changes in file: bitcoingui.cpp<br>
→ Changing the size of the icons in "STATUSBAR_ICONSIZE" from 16 to 64 - Changes in file: guiconstants.h<br>
→ Adding LightBurden.conf file opening from the Console Window menu<br>
→ Changed description of mining type PoW<br>
→ Changed description of mining type PoS<br>
→ New PoS Mining Icon<br>
→ Changed version to 2.6.4.5<br>

>$0022 2018/12/08<br>
→ Alterada Versão para 2.6.4.6;<br>
→ Alterada versão do protocolo para mínimo 700001;<br>
→ Adicionada Introdução;<br>
→ Adicionado pontos de checagem de blocos até o bloco 180000<br>

>$0028 2019/05/31<br>
→ Alterada Versão para 2.6.4.7;<br>
→ Alterada versão do protocolo mínimo para 700002;<br>
→ Adicionada página Charity(Doações POS);<br>
→ Adicionados endereços Donate Foudation;<br>
→ Correção das strings que determinam as fases de mineração;<br>
→ Preparação para o ambiente de mineração PoW após o bloco 263250;<br>
→ Mudança dos ícones e logos;<br>
→ Adição do parâmetro -zapwallettxes;<br>

> 2020/10/22<br>
→ [UPDATE] Nova versão 2.6.5.0;<br>
→ [UPDATE] Adição de lista de endereços de participantes do sistema de doações(CHARITY);<br>
→ [OPMIZATION] Limpeza e Remoção de arquivos desnecessários da base do código;<br>
→ [FEATURE] Compatibilidade com Openssl versões 1.0.2 e acima;<br>
→ [FEATURE] Compatibilidade com Boost 1.66 e 1.70;<br>
→ [FEATURE] Implementação de auto-completar na janela de depuração;<br>
→ [FIX] getbalance() bug;<br>
→ [FIX] TransactionView: Cálculo do último mês;<br>
→ [UPDATE] Array de seed nodes;<br>
→ [UPDATE] Checkpoints;<br>
→ [UPDATE] Arquivos de Compilação LightBurden-qt.pro e makefile.mingw usados em compilação Windows<br>

> 2021/01/29<br>
→ [UPDATE] Nova versão 2.6.5.1;<br>
→ [UPDATE] Alterada versão do protocolo para mínimo 700003;<br>
→ [GUI] Novo Design QT;<br>
→ [GUI] Nova Splash Screen;<br>
→ [FIX] Dev fee;<br>
→ [FIX] Staking Dust;<br>
→ [FIX] Compilações MinGW/QT;<br>
→ [UPDATE] Novas premiações PoW e PoS;<br>
→ [UPDATE] Novos endereços Foundation PoS;<br>
→ [NEW] Halving Blocks<br>

> 2021/03/18<br>
→ [NEW] Checkpoints;<br>
→ [FIX] Mineração PoS;<br>
→ [BUMP] Alterada versão do protocolo mínimo para 700004;<br>
→ [BUMP] Nova versão 2.6.5.2;<br>

> 2021/10/14<br>
→ [NEW] Checkpoints;<br>
→ [REMOVE] Pagamentos para a Foundation PoS;<br>
→ [NEW] Fase especial de mineração para a Foundation PoW;<br>
→ [FORK] Preparação do código para o hardfork;<br>
→ [BUMP] Alterada versão do protocolo mínimo para 700006;<br>
→ [BUMP] Nova versão 2.6.5.4;<br>

src/base58.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2012 The Bitcoin Developers
// Copyright (c) 2022 The LightBurden Developers
// Distributed under the MIT/X11 software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.


//
// Why base-58 instead of standard base-64 encoding?
// - Don't want 0OIl characters that look the same in some fonts and
//      could be used to create visually identical looking account numbers.
// - A string with non-alphanumeric characters is not as easily accepted as an account number.
// - E-mail usually won't line-break if there's no punctuation to break at.
// - Double-clicking selects the whole number as one word if it's all alphanumeric.
//
#ifndef BITCOIN_BASE58_H
#define BITCOIN_BASE58_H

#include <string>
#include <vector>
#include "bignum.h"
#include "key.h"
#include "script.h"

static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

// Encode a byte sequence as a base58-encoded string
inline std::string EncodeBase58(const unsigned char* pbegin, const unsigned char* pend)
{
    CAutoBN_CTX pctx;
    CBigNum bn58 = 58;
    CBigNum bn0 = 0;

    // Convert big endian data to little endian
    // Extra zero at the end make sure bignum will interpret as a positive number
    std::vector<unsigned char> vchTmp(pend-pbegin+1, 0);
    reverse_copy(pbegin, pend, vchTmp.begin());

    // Convert little endian data to bignum
    CBigNum bn;
    bn.setvch(vchTmp);

    // Convert bignum to std::string
    std::string str;
    // Expected size increase from base58 conversion is approximately 137%
    // use 138% to be safe
    str.reserve((pend - pbegin) * 138 / 100 + 1);
    CBigNum dv;
    CBigNum rem;
    while (bn > bn0)
    {
#if OPENSSL_VERSION_NUMBER < 0x10100000
        if (!BN_div(&dv, &rem, &bn, &bn58, pctx))
#else
        if (!BN_div(dv.pbn, rem.pbn, bn.pbn, bn58.pbn, pctx))
#endif
            throw bignum_error("EncodeBase58 : BN_div failed");
        bn = dv;
        unsigned int c = rem.getulong();
        str += pszBase58[c];
    }

    // Leading zeroes encoded as base58 zeros
    for (const unsigned char* p = pbegin; p < pend && *p == 0; p++)
        str += pszBase58[0];

    // Convert little endian std::string to big endian
    reverse(str.begin(), str.end());
    return str;
}

// Encode a byte vector as a base58-encoded string
inline std::string EncodeBase58(const std::vector<unsigned char>& vch)
{
    return EncodeBase58(&vch[0], &vch[0] + vch.size());
}

// Decode a base58-encoded string psz into byte vector vchRet
// returns true if decoding is successful
inline bool DecodeBase58(const char* psz, std::vector<unsigned char>& vchRet)
{
    CAutoBN_CTX pctx;
    vchRet.clear();
    CBigNum bn58 = 58;
    CBigNum bn = 0;
    CBigNum bnChar;
    while (isspace(*psz))
        psz++;

    // Convert big endian string to bignum
    for (const char* p = psz; *p; p++)
    {
        const char* p1 = strchr(pszBase58, *p);
        if (p1 == NULL)
        {
            while (isspace(*p))
                p++;
            if (*p != '\0')
                return false;
            break;
        }
        bnChar.setulong(p1 - pszBase58);
#if OPENSSL_VERSION_NUMBER < 0x10100000
        if (!BN_mul(&bn, &bn, &bn58, pctx))
#else
        if (!BN_mul(bn.pbn, bn.pbn, bn58.pbn, pctx))
#endif
            throw bignum_error("DecodeBase58 : BN_mul failed");
        bn += bnChar;
    }

    // Get bignum as little endian data
    std::vector<unsigned char> vchTmp = bn.getvch();

    // Trim off sign byte if present
    if (vchTmp.size() >= 2 && vchTmp.end()[-1] == 0 && vchTmp.end()[-2] >= 0x80)
        vchTmp.erase(vchTmp.end()-1);

    // Restore leading zeros
    int nLeadingZeros = 0;
    for (const char* p = psz; *p == pszBase58[0]; p++)
        nLeadingZeros++;
    vchRet.assign(nLeadingZeros + vchTmp.size(), 0);

    // Convert little endian data to big endian
    reverse_copy(vchTmp.begin(), vchTmp.end(), vchRet.end() - vchTmp.size());
    return true;
}

// Decode a base58-encoded string str into byte vector vchRet
// returns true if decoding is successful
inline bool DecodeBase58(const std::string& str, std::vector<unsigned char>& vchRet)
{
    return DecodeBase58(str.c_str(), vchRet);
}




// Encode a byte vector to a base58-encoded string, including checksum
inline std::string EncodeBase58Check(const std::vector<unsigned char>& vchIn)
{
    // add 4-byte hash check to the end
    std::vector<unsigned char> vch(vchIn);
    uint256 hash = Hash(vch.begin(), vch.end());
    vch.insert(vch.end(), (unsigned char*)&hash, (unsigned char*)&hash + 4);
    return EncodeBase58(vch);
}

// Decode a base58-encoded string psz that includes a checksum, into byte vector vchRet
// returns true if decoding is successful
inline bool DecodeBase58Check(const char* psz, std::vector<unsigned char>& vchRet)
{
    if (!DecodeBase58(psz, vchRet))
        return false;
    if (vchRet.size() < 4)
    {
        vchRet.clear();
        return false;
    }
    uint256 hash = Hash(vchRet.begin(), vchRet.end()-4);
    if (memcmp(&hash, &vchRet.end()[-4], 4) != 0)
    {
        vchRet.clear();
        return false;
    }
    vchRet.resize(vchRet.size()-4);
    return true;
}

// Decode a base58-encoded string str that includes a checksum, into byte vector vchRet
// returns true if decoding is successful
inline bool DecodeBase58Check(const std::string& str, std::vector<unsigned char>& vchRet)
{
    return DecodeBase58Check(str.c_str(), vchRet);
}





/** Base class for all base58-encoded data */
class CBase58Data
{
protected:
    // the version byte
    unsigned char nVersion;

    // the actually encoded data
    std::vector<unsigned char> vchData;

    CBase58Data()
    {
        nVersion = 0;
        vchData.clear();
    }

    ~CBase58Data()
    {
        // zero the memory, as it may contain sensitive data
        if (!vchData.empty())
            memset(&vchData[0], 0, vchData.size());
    }

    void SetData(int nVersionIn, const void* pdata, size_t nSize)
    {
        nVersion = nVersionIn;
        vchData.resize(nSize);
        if (!vchData.empty())
            memcpy(&vchData[0], pdata, nSize);
    }

    void SetData(int nVersionIn, const unsigned char *pbegin, const unsigned char *pend)
    {
        SetData(nVersionIn, (void*)pbegin, pend - pbegin);
    }

public:
    bool SetString(const char* psz)
    {
        std::vector<unsigned char> vchTemp;
        DecodeBase58Check(psz, vchTemp);
        if (vchTemp.empty())
        {
            vchData.clear();
            nVersion = 0;
            return false;
        }
        nVersion = vchTemp[0];
        vchData.resize(vchTemp.size() - 1);
        if (!vchData.empty())
            memcpy(&vchData[0], &vchTemp[1], vchData.size());
        memset(&vchTemp[0], 0, vchTemp.size());
#if OPENSSL_VERSION_NUMBER >= 0x10100000        
        OPENSSL_cleanse(&vchTemp[0], vchData.size());
#endif        
        return true;
    }

    bool SetString(const std::string& str)
    {
        return SetString(str.c_str());
    }

    std::string ToString() const
    {
        std::vector<unsigned char> vch(1, nVersion);
        vch.insert(vch.end(), vchData.begin(), vchData.end());
        return EncodeBase58Check(vch);
    }

    int CompareTo(const CBase58Data& b58) const
    {
        if (nVersion < b58.nVersion) return -1;
        if (nVersion > b58.nVersion) return  1;
        if (vchData < b58.vchData)   return -1;
        if (vchData > b58.vchData)   return  1;
        return 0;
    }

    bool operator==(const CBase58Data& b58) const { return CompareTo(b58) == 0; }
    bool operator<=(const CBase58Data& b58) const { return CompareTo(b58) <= 0; }
    bool operator>=(const CBase58Data& b58) const { return CompareTo(b58) >= 0; }
    bool operator< (const CBase58Data& b58) const { return CompareTo(b58) <  0; }
    bool operator> (const CBase58Data& b58) const { return CompareTo(b58) >  0; }
};

/** base58-encoded addresses.
 * Public-key-hash-addresses have version 25 (or 111 testnet).
 * The data vector contains RIPEMD160(SHA256(pubkey)), where pubkey is the serialized public key.
 * Script-hash-addresses have version 85 (or 196 testnet).
 * The data vector contains RIPEMD160(SHA256(cscript)), where cscript is the serialized redemption script.
 */
class CBitcoinAddress;
class CBitcoinAddressVisitor : public boost::static_visitor<bool>
{
private:
    CBitcoinAddress *addr;
public:
    CBitcoinAddressVisitor(CBitcoinAddress *addrIn) : addr(addrIn) { }
    bool operator()(const CKeyID &id) const;
    bool operator()(const CScriptID &id) const;
    bool operator()(const CNoDestination &no) const;
};

class CBitcoinAddress : public CBase58Data
{
public:
    enum
    {
        PUBKEY_ADDRESS = 63, // Letra inicial dos endereços( 63 = S )
        SCRIPT_ADDRESS = 5,
        PUBKEY_ADDRESS_TEST = 125, // Letra inicial dos endereços( 125 = s ) - TESTNET
        SCRIPT_ADDRESS_TEST = 196,
    };

    bool Set(const CKeyID &id) {
        SetData(fTestNet ? PUBKEY_ADDRESS_TEST : PUBKEY_ADDRESS, &id, 20);
        return true;
    }

    bool Set(const CScriptID &id) {
        SetData(fTestNet ? SCRIPT_ADDRESS_TEST : SCRIPT_ADDRESS, &id, 20);
        return true;
    }

    bool Set(const CTxDestination &dest)
    {
        return boost::apply_visitor(CBitcoinAddressVisitor(this), dest);
    }

    bool IsValid() const
    {
        unsigned int nExpectedSize = 20;
        bool fExpectTestNet = false;
        switch(nVersion)
        {
            case PUBKEY_ADDRESS:
                nExpectedSize = 20; // Hash of public key
                fExpectTestNet = false;
                break;
            case SCRIPT_ADDRESS:
                nExpectedSize = 20; // Hash of CScript
                fExpectTestNet = false;
                break;

            case PUBKEY_ADDRESS_TEST:
                nExpectedSize = 20;
                fExpectTestNet = true;
                break;
            case SCRIPT_ADDRESS_TEST:
                nExpectedSize = 20;
                fExpectTestNet = true;
                break;

            default:
                return false;
        }
        return fExpectTestNet == fTestNet && vchData.size() == nExpectedSize;
    }

    CBitcoinAddress()
    {
    }

    CBitcoinAddress(const CTxDestination &dest)
    {
        Set(dest);
    }

    CBitcoinAddress(const std::string& strAddress)
    {
        SetString(strAddress);
    }

    CBitcoinAddress(const char* pszAddress)
    {
        SetString(pszAddress);
    }

    CTxDestination Get() const {
        if (!IsValid())
            return CNoDestination();
        switch (nVersion) {
        case PUBKEY_ADDRESS:
        case PUBKEY_ADDRESS_TEST: {
            uint160 id;
            memcpy(&id, &vchData[0], 20);
            return CKeyID(id);
        }
        case SCRIPT_ADDRESS:
        case SCRIPT_ADDRESS_TEST: {
            uint160 id;
            memcpy(&id, &vchData[0], 20);
            return CScriptID(id);
        }
        }
        return CNoDestination();
    }

    bool GetKeyID(CKeyID &keyID) const {
        if (!IsValid())
            return false;
        switch (nVersion) {
        case PUBKEY_ADDRESS:
        case PUBKEY_ADDRESS_TEST: {
            uint160 id;
            memcpy(&id, &vchData[0], 20);
            keyID = CKeyID(id);
            return true;
        }
        default: return false;
        }
    }

    bool IsScript() const {
        if (!IsValid())
            return false;
        switch (nVersion) {
        case SCRIPT_ADDRESS:
        case SCRIPT_ADDRESS_TEST: {
            return true;
        }
        default: return false;
        }
    }
};

bool inline CBitcoinAddressVisitor::operator()(const CKeyID &id) const         { return addr->Set(id); }
bool inline CBitcoinAddressVisitor::operator()(const CScriptID &id) const      { return addr->Set(id); }
bool inline CBitcoinAddressVisitor::operator()(const CNoDestination &id) const { return false; }

/** A base58-encoded secret key */
class CBitcoinSecret : public CBase58Data
{
public:
    void SetSecret(const CSecret& vchSecret, bool fCompressed)
    {
        assert(vchSecret.size() == 32);
        SetData(128 + (fTestNet ? CBitcoinAddress::PUBKEY_ADDRESS_TEST : CBitcoinAddress::PUBKEY_ADDRESS), &vchSecret[0], vchSecret.size());
        if (fCompressed)
            vchData.push_back(1);
    }

    CSecret GetSecret(bool &fCompressedOut)
    {
        CSecret vchSecret;
        vchSecret.resize(32);
        memcpy(&vchSecret[0], &vchData[0], 32);
        fCompressedOut = vchData.size() == 33;
        return vchSecret;
    }

    bool IsValid() const
    {
        bool fExpectTestNet = false;
        switch(nVersion)
        {
            case (128 + CBitcoinAddress::PUBKEY_ADDRESS):
                break;

            case (128 + CBitcoinAddress::PUBKEY_ADDRESS_TEST):
                fExpectTestNet = true;
                break;

            default:
                return false;
        }
        return fExpectTestNet == fTestNet && (vchData.size() == 32 || (vchData.size() == 33 && vchData[32] == 1));
    }

    bool SetString(const char* pszSecret)
    {
        return CBase58Data::SetString(pszSecret) && IsValid();
    }

    bool SetString(const std::string& strSecret)
    {
        return SetString(strSecret.c_str());
    }

    CBitcoinSecret(const CSecret& vchSecret, bool fCompressed)
    {
        SetSecret(vchSecret, fCompressed);
    }

    CBitcoinSecret()
    {
    }
};

#endif
